lsat/store.go

```diff
@@ -63,31 +69,83 @@ func (f *FileStore) HasToken() bool {
 	return fileExists(f.fileName)
 }
 
-// Token returns the token that is contained in the store or an error if there
-// is none.
+// CurrentToken returns the token that is currently contained in the store or an
+// error if there is none.
 //
 // NOTE: This is part of the Store interface.
-func (f *FileStore) Token() (*Token, error) {
+func (f *FileStore) CurrentToken() (*Token, error) {
 	if !f.HasToken() {
 		return nil, ErrNoToken
 	}
-	bytes, err := ioutil.ReadFile(f.fileName)
+	return readTokenFile(f.fileName)
+}
+
+// AllTokens returns all tokens that the store has knowledge of, even
+// if they might be expired.
+func (f *FileStore) AllTokens() ([]*Token, error) {
+	tokens := make([]*Token, 0)
+
+	// All tokens start with the same name but old tokens have a timestamp
+	// appended to the original file name.
+	tokenDir := filepath.Dir(f.fileName)
+	files, err := ioutil.ReadDir(tokenDir)
 	if err != nil {
 		return nil, err
 	}
-	return deserializeToken(bytes)
+	for _, file := range files {
+		name := file.Name()
+		if !strings.HasPrefix(name, storeFileName) {
+			continue
+		}
+		token, err := readTokenFile(filepath.Join(tokenDir, name))
+		if err != nil {
+			return nil, err
+		}
+		tokens = append(tokens, token)
+	}
+
+	return tokens, nil
 }
 
 // StoreToken saves a token to the store, overwriting any old token if there is
 // one.
 //
 // NOTE: This is part of the Store interface.
 func (f *FileStore) StoreToken(token *Token) error {
+	// Serialize the token first, before we rename anything.
 	bytes, err := serializeToken(token)
 	if err != nil {
 		return err
 	}
-	return ioutil.WriteFile(f.fileName, bytes, 0600)
+
+	// Now try renaming the current file to a new file name with a timestamp
+	// that we keep around for accounting purposes.
+	var newFileName string
+	if f.HasToken() {
+		timestampSuffix := time.Now().Format("2006-01-02T15-04-05")
+		newFileName = f.fileName + timestampSuffix
+		err := os.Rename(f.fileName, newFileName)
+		if err != nil {
+			return err
+		}
+	}
+
+	// We want this operation to be atomic. So if the write of the new file
+	// fails, we want to move the old one back into its position.
+	err = ioutil.WriteFile(f.fileName, bytes, 0600)
+	if err != nil && newFileName != "" {
+		_ = os.Rename(newFileName, f.fileName)
```

Comment:
Following my other comment, I think we can actually just return the error here and not rename the file back (what you had previously) because we'll have a pending state that we can recover from for the new token we're attempting to store.

Reply:
If the write of the new file fails, we do want to move the old file (which contained the pending state) back, otherwise the store won't return true on `HasToken()` the next time. If the rename back fails, we _should_ return the error though, because then we really are in a strange state.